File: Algorithm/338. Counting Bits.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-07-19 15:19:08
# @Last modified by:   WuLC
# @Last Modified time: 2016-07-19 15:19:21
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
        result = [0 for i in xrange(num+1)]
        for i in xrange(1, num+1):
            result[i] = result[i>>1] + (i&1)
        return result

File: Algorithm/Java/338. Counting Bits.java
/*
* @Author: WuLC
* @Date:   2017-05-17 12:37:18
* @Last Modified by:   WuLC
* @Last Modified time: 2017-05-17 12:42:57
* @Email: liangchaowu5@gmail.com
*/

// bit manipulation and dp, time O(n)
// important rule: result[i] = result[i>>1] + (i&1);
public class Solution 
{
    public int[] countBits(int num) 
    {
        int[] result = new int[num+1];
        for(int i = 0; i <= num; i++)
            result[i] = result[i>>1] + (i&1);
        return result;
    }
}